runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java

```diff
@@ -629,6 +662,16 @@ public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exc
                 ctx.close();
             } else if (e.state() == IdleState.WRITER_IDLE) {
                 keepAlive();
+            } else if (evt instanceof SslHandshakeCompletionEvent) {
+                // Event fired once the SSL handshake is complete, which may have succeeded or not.
+                // If SSL handshake failed, activate flag to exponentially backoff reconnection time.
+                SslHandshakeCompletionEvent sslEvt = (SslHandshakeCompletionEvent) evt;
+                if (sslEvt.cause() instanceof SSLHandshakeException) {
```

Comment:
I don't think this fix should be specific to ssl handshake failures. We should should solve the general problem (i.e. clients spinning on errors). Any exception can cause the channel to be closed, for example:
1. client/server handshake 
2. authentication failures
3. encode/decode failures
4. other errors

In all of those cases, we should have an exponential backoff.

Reply:
Good points! Is your suggestion to handle all these as part of this PR or create issues that could be added to our backlog.

Reply:
From the logs I could observe these exceptions are wrapped by the SSLHandshakeException.

Reply:
I don't think these cases are handled differently. We just need to apply the exponential backoff to the reconnect logic.

Reply:
> From the logs I could observe these exceptions are wrapped by the SSLHandshakeException.

Why endcode/decode failures are wrapped by SshHandshakeException?

Reply:
When an exception happens during encode/decode we close the channel and disconnect, then subsequent requests will try to open a new channel thus triggering the reconnect logic.